system-helper/flatpak-system-helper.c

```diff
@@ -51,6 +59,236 @@ G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitAuthorizationResult, g_object_unref)
+  GList *l;
+  gboolean mount_successful = FALSE;
+
+  mounts = g_unix_mounts_get (NULL);
+  for (l = mounts; l != NULL; l = l->next)
+    {
+      mount_entry = (GUnixMountEntry *) l->data;
+      if (g_strcmp0 (pull->user_dir_name, g_unix_mount_get_mount_path (mount_entry)) == 0)
+        mount_successful = TRUE;
+    }
+  g_list_free_full (mounts, (GDestroyNotify) g_unix_mount_free);
+
+  if (mount_successful)
+    {
+      g_autofree gchar *basename = NULL;
+      g_autofree gchar *child_repo_path = NULL;
+
+      g_signal_handler_disconnect (monitor, pull->mount_monitor_id);
+      pull->mount_monitor_id = 0;
+
+      basename = g_path_get_basename (pull->child_repo_path);
+      child_repo_path = g_build_filename (pull->user_dir_name, basename, NULL);
+
+      flatpak_system_helper_complete_get_child_repo_for_pull (pull->object,
+                                                              pull->invocation,
+                                                              child_repo_path);
+
+      g_source_remove (pull->unsuccessful_mount_timeout_id);
+      pull->unsuccessful_mount_timeout_id = 0;
+    }
+}
+
+static gboolean
+unsuccessful_mount_timeout (gpointer user_data)
+{
+  FlatpakSystemHelperOngoingPull *pull = user_data;
+
+  system_helper_ongoing_pull_teardown (pull);
+  g_dbus_method_invocation_return_error (pull->invocation, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                                         "Failed to mount %s on %s with bindfs",
+                                         pull->user_dir_name, pull->root_dir_name);
+
+  return FALSE;
+}
+
+static FlatpakSystemHelperOngoingPull *
+system_helper_ongoing_pull_new (FlatpakSystemHelper   *object,
+                                GDBusMethodInvocation *invocation,
+                                const gchar           *ref,
+                                const gchar           *commit,
+                                uid_t                  uid,
+                                gchar                 *user_dir_name,
+                                gchar                 *root_dir_name,
+                                gchar                 *child_repo_path)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  FlatpakSystemHelperOngoingPull *pull;
+
+  pull = g_slice_new0 (FlatpakSystemHelperOngoingPull);
+  pull->object = object;
+  pull->invocation = invocation;
+  pull->ref = g_strdup (ref);
+  pull->commit = g_strdup (commit);
+  pull->cancellable = g_cancellable_new ();
+  pull->uid = uid;
+  pull->ref_count = 1;
+
+  pull->watch_id = g_bus_watch_name_on_connection (connection,
+                                                   g_dbus_connection_get_unique_name (connection),
+                                                   G_BUS_NAME_WATCHER_FLAGS_NONE, NULL,
+                                                   name_vanished_callback,
+                                                   system_helper_ongoing_pull_ref (pull), NULL);
```

Comment:
Isn't this keeping the OngoingPull struct alive until the caller disappears ? Is there some code other than the name_vanished_callback that clears the ref you take here ?

It might not matter for the cli, but gnome-software and other clients can be long-living...

Reply:
It’s keeping the `OngoingPull` struct alive until the name watch is removed, which happens in `system_helper_ongoing_pull_teardown()`. That’s guaranteed to be called before the final ref to the `OngoingPull` struct is removed, so no ref should be needed here. In other words, the `OngoingPull` struct should be alive and valid for the duration of the name watch. To ref here would introduce a refcount loop.